src/monodroid/monodroid.targets

```diff
@@ -48,56 +21,97 @@
-        DestinationFiles="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.debug.d.so')"
+  </Target>
+
+  <Target Name="_ConfigureAndroidRuntimes"
+      Inputs="CMakeLists.txt"
+      Outputs="@(AndroidSupportedTargetJitAbi->'$(IntermediateOutputPath)\%(Identity)-Debug\CMakeCache.txt');@(AndroidSupportedTargetJitAbi->'$(IntermediateOutputPath)\%(Identity)-Release\CMakeCache.txt')">
+    <MakeDir Directories="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.Identity)-Debug" />
+    <Exec
+        Command="$(CmakePath) $(_CmakeAndroidFlags) -DCONFIGURATION=Release -DCMAKE_BUILD_TYPE=Debug -DANDROID_ABI=%(AndroidSupportedTargetJitAbi.Identity) -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=$(OutputPath)/%(AndroidSupportedTargetJitAbi.Identity) $(MSBuildThisFileDirectory)"
+        WorkingDirectory="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.Identity)-Debug"
     />
-    <Copy
-        SourceFiles="@(AndroidSupportedTargetJitAbi->'libs\Debug\%(Identity)\libmonodroid.so')"
-        DestinationFiles="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.debug.so')"
+
+    <MakeDir Directories="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.Identity)-Release" />
+    <Exec
+        Command="$(CmakePath) $(_CmakeAndroidFlags) -DSTRIP_DEBUG=ON -DCONFIGURATION=Debug -DCMAKE_BUILD_TYPE=Release -DANDROID_ABI=%(AndroidSupportedTargetJitAbi.Identity) -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=$(OutputPath)/%(AndroidSupportedTargetJitAbi.Identity) $(MSBuildThisFileDirectory)"
+        WorkingDirectory="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.Identity)-Release"
     />
-    <WriteLinesToFile
-        File="jni\Application.mk"
-        Lines="APP_ABI := $(_AppAbi); APP_CFLAGS+=-DRELEASE=1"
-        Overwrite="True"
+  </Target>
+
+  <Target Name="_BuildAndroidRuntimes"
+      DependsOnTargets="_ConfigureAndroidRuntimes"
+      Inputs="@(AndroidSupportedTargetJitAbi->'$(IntermediateOutputPath)\%(Identity)-Debug\CMakeCache.txt');@(AndroidSupportedTargetJitAbi->'$(IntermediateOutputPath)\%(Identity)-Release\CMakeCache.txt')"
+      Outputs="@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.debug.so');@(AndroidSupportedTargetJitAbi->'$(OutputPath)\%(Identity)\libmono-android.release.so')">
+    <Exec
+        Command="$(NinjaPath) -v"
+        WorkingDirectory="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.Identity)-Debug"
     />
+
+    <Exec
+        Command="$(NinjaPath) -v"
+        WorkingDirectory="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.Identity)-Release"
+    />
+  </Target>
+
+  <Target Name="_BuildHostRuntimes"
+      DependsOnTargets="_GetBuildHostRuntimes;_CreateJavaInteropDllConfig;_ConfigureHostRuntimes"
+      Inputs="@(_HostRuntime->'$(IntermediatePath)%(OutputDirectory)-Debug\CMakeCache.txt');@(_HostRuntime->'$(IntermediatePath)%(OutputDirectory)-Release\CMakeCache.txt')"
```

Comment:
Do we really want to make the `_BuildAndroidRuntimes` target conditional on *only* `CMakeCache.txt` (in turn dependent on `CMakeLists.txt`)?

This means that any changes to `.c` files won't trigger a rebuild.  That seems bad.

Perhaps we should bring back the `@(_CFile)`/`$(_UnixCFile)`?

Reply:
I don't want to have two lists of .c (and in the future .cpp) files in two different places... About the only two ideas I have are 

  * use `**.c` in `Inputs`
  * don't depend on `CMakeCache.txt` and run the target unconditionally, letting cmake figure things out

However, I don't like either of them, not really. Perhaps there's a way for cmake to generate the list of sources and put it in some file which can be included in the .csproj - that way the first build will always run, further builds will use the include.